File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> parentheses_result;
        queue<pair<string, int>> parentheses;
        int len = 1;

        parentheses.push({"(", 1});

        while(!parentheses.empty()){
            int size = parentheses.size();

            if(len == 2*n){
                while(!parentheses.empty()){
                    parentheses_result.push_back(parentheses.front().first);
                    parentheses.pop();
                }
                break;
            }

            len += 1;
            for(int i = 0; i < size; i++){
                const auto cur = parentheses.front();
                parentheses.pop();

                int right = cur.first.size() - cur.second;
                if(cur.second > right){
                    parentheses.push({cur.first+")", cur.second});
                }
                if(cur.second < n){
                    parentheses.push({cur.first+"(", cur.second+1});
                }
                
            }
        }
        return parentheses_result;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # maintain left num and right number

        def backtrack(cur, left, right, result):
            if len(cur) == 2 * n:
                result.append("".join(cur))
                return

            if left < n:
                cur.append("(")
                backtrack(cur, left + 1, right, result)
                cur.pop()

            if left > right:
                cur.append(")")
                backtrack(cur, left , right + 1, result)
                cur.pop()

        result = []
        backtrack([], 0, 0, result)


        return result
            
        
        